fix(ast)!: add `TSGlobalDeclaration` type

overlookmotel opened the pull request:
Fix #14860.

### The problem

`TSModuleDeclaration` type was used to represent 4 different syntaxes:

* `namespace X {}`
* `module X {}`
* `module "x" {}`
* `global {}`

(and all the above prefixed with `declare`)

`global {}` is quite different from the others in a few ways:

* It has no identifier (`global` is a keyword, not an identifier).
* It has no symbol.
* The `body` block is not optional.
* Directives are not allowed in the `body` block.
* It is always ambient context (regardless of whether it has `declare` modifier).

In particular, it was causing panics because `symbol_id` field for `TSModuleDeclaration`s representing `global {}` was `None`.

### Solution

Rather than trying to make `TSModuleDeclaration` type encode all these differences, and make Oxc's code reason about them, split `global {}` into a new type `TSGlobalDeclaration`.

The rest of the diff is just adjusting code to also handle `TSGlobalDeclaration` where it was dealing with `TSModuleDeclaration`s.

This change simplifies the parser quite a bit.

`oxc_semantic`'s `SymbolDeclarationTest` tester now tests that all `TSModuleDeclaration`s with `id: BindingIdentifier` do have a symbol:

https://github.com/oxc-project/oxc/pull/15712/files#diff-ce544d71b839faa898887ac920237e563e59a708e7fca86b53023498ab67a611L101-R110

overlookmotel commented on crates/oxc_transformer/src/typescript/namespace.rs:
```diff
@@ -55,8 +55,15 @@ impl<'a> Traverse<'a, TransformState<'a>> for TypeScriptNamespace<'a, '_> {
+                }
                 Statement::ExportNamedDeclaration(export_decl)
                     if export_decl.declaration.as_ref().is_some_and(|declaration| {
+                        // Note: No need to check for `TSGlobalDeclaration` here, as it can't be exported
```
Good idea! Done in 3f85660b073767dfc9722b282f2b6f7f63a48dbf.

overlookmotel commented:
https://github.com/oxc-project/monitor-oxc/actions/runs/19389800278

overlookmotel commented:
> https://github.com/oxc-project/monitor-oxc/actions/runs/19389800278

I don't know how to interpret these results. Everything is also broken on main, right?

Boshen reviewed (dismissed):
Already queued a few breaking changes in today's release, so block merge until tomorrow.

Dunqing reviewed (approved):
I've fixed the formatter issue; feel free to merge it!

overlookmotel commented on crates/oxc_transformer/src/typescript/namespace.rs:
```diff
@@ -198,6 +206,10 @@ impl<'a> TypeScriptNamespace<'a, '_> {
+                Statement::TSGlobalDeclaration(_) => {
+                    // Remove it
+                    // TODO: Is it even possible for a`TSGlobalDeclaration` to be nested within a `TSModuleDeclaration`?
+                }
```
You're right. But I want to get this merged, so I'll address it in a follow-up PR.

overlookmotel committed:
fix(ast)!: add `TSGlobalDeclaration` type (#15712)

Fix #14860.

### The problem

`TSModuleDeclaration` type was used to represent 4 different syntaxes:

* `namespace X {}`
* `module X {}`
* `module "x" {}`
* `global {}`

(and all the above prefixed with `declare`)

`global {}` is quite different from the others in a few ways:

* It has no identifier (`global` is a keyword, not an identifier).
* It has no symbol.
* The `body` block is not optional.
* Directives are not allowed in the `body` block.
* It is always ambient context (regardless of whether it has `declare` modifier).

In particular, it was causing panics because `symbol_id` field for `TSModuleDeclaration`s representing `global {}` was `None`.

### Solution

Rather than trying to make `TSModuleDeclaration` type encode all these differences, and make Oxc's code reason about them, split `global {}` into a new type `TSGlobalDeclaration`.

The rest of the diff is just adjusting code to also handle `TSGlobalDeclaration` where it was dealing with `TSModuleDeclaration`s.

This change simplifies the parser quite a bit.

`oxc_semantic`'s `SymbolDeclarationTest` tester now tests that all `TSModuleDeclaration`s with `id: BindingIdentifier` do have a symbol:

https://github.com/oxc-project/oxc/pull/15712/files#diff-ce544d71b839faa898887ac920237e563e59a708e7fca86b53023498ab67a611L101-R110

### Formatter bug

There is one regression in the formatter for `typescript/module/global.ts` case.

https://github.com/oxc-project/oxc/pull/15712/files#diff-f248b1d49b8c482864810055632d7322cdebc895dd4dfe05fbefcf5e23ba12c9R51

Input:

```ts
declare /* module */ global {}
declare /* namespace */ global {}
```

Output:

```diff
- declare /* module */ global {}
- declare /* namespace */ global {}
+ declare global /* module */ {}
+ declare global /* namespace */ {}
```

I assume this is happening because there's no longer a `BindingIdentifier` for the `global` keyword for the comment to get attached to.

I am not sure how to fix this. Maybe the `global_span` field can help?

@Dunqing This PR touches a *lot* of files. Can I propose that we merge it before it gets conflicts, and if you can give me some guidance, I'll do a follow-up PR to fix it?
